Optimize dead private global variables

zoecarver opened the pull request:
<!-- What's in this pull request? -->
This optimization removes private global variables when the only use is an allocation and a store instruction. To make sure I didn't remove used variables, I updated `SILGlobalOpt` to collect instructions from  _all_ functions. Then, it bails out when trying to optimize, if it finds an instruction with a parent function that shouldn't be optimized. 

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-6597](https://bugs.swift.org/browse/SR-6597).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

zoecarver committed:
Add optimization to remove unused global private variables

zoecarver committed:
Refactor and add tests

zoecarver commented:
cc @eeckstein

slavapestov commented on lib/SILOptimizer/IPO/GlobalOpt.cpp:
```diff
@@ -770,6 +786,35 @@ void SILGlobalOpt::optimizeInitializer(SILFunction *AddrF,
+/// and the alloc. Otherwise, return.
+void SILGlobalOpt::tryOptimizeUnusedGlobal(SILGlobalVariable *global,
+                                           StoreInst *store) {
+  if (global->getLinkage() != SILLinkage::Private)
```
I think the canonical check is `isPossiblyUsedExternally(global->getLinkage(), module->isWholeModule())`, which should also handle hidden linkage in WMO builds

slavapestov commented on test/SILOptimizer/remove_unused_global_vars.swift:
```diff
@@ -0,0 +1,28 @@
+// CHECK: sil_global @${{.*}}unused4{{.*}} : $Int
+public var unused4 = 0
+
+// These should only be optimized with -wmo.
```
Does this already work? I would expect they get Hidden linkage not Private and wouldn't be caught by the check above.

zoecarver replied:
It didn't use to work. After applying your above comment it now works. I'll add tests for this too.

slavapestov replied:
Great!

slavapestov commented on test/SILOptimizer/remove_unused_global_vars.swift:
```diff
@@ -0,0 +1,28 @@
+// CHECK-NOT: sil_global private [let] {{.*}}unused1{{.*}}
+private let unused1 = 0
+// CHECK-NOT: sil_global private {{.*}}unused2{{.*}}
+private var unused2 = 42
```
Can you add some tests where the global has a non-trivial initial value that doesn't get optimized down to a constant? We should be able to eliminate the accessor function as well.

zoecarver replied:
Added the tests you suggested.

slavapestov commented on lib/SILOptimizer/IPO/GlobalOpt.cpp:
```diff
@@ -770,6 +786,35 @@ void SILGlobalOpt::optimizeInitializer(SILFunction *AddrF,
+                                           StoreInst *store) {
+  if (global->getLinkage() != SILLinkage::Private)
+    return;
+  if (GlobalLoadMap.count(global) || GlobalAccessMap.count(global))
```
It might be safer to look for any use of the global except for a store, not just load and begin_access.

Also if the begin_access is for a store, you could still potentially remove the global.

zoecarver replied:
Is there a way to get all the uses of a `SILGlobalVariable`? It doesn't look like it.

slavapestov replied:
You're right, there's no way of doing that right now, so handling specific instructions that directly reference SILGlobalVariables is your best bet.

slavapestov commented on lib/SILOptimizer/IPO/GlobalOpt.cpp:
```diff
@@ -770,6 +786,35 @@ void SILGlobalOpt::optimizeInitializer(SILFunction *AddrF,
+
+  auto globalAddr = cast<GlobalAddrInst>(store->getDest());
+  bool canRemove = true;
+  for (auto *use : globalAddr->getUses()) {
```
This loop only looks at uses within this function, so if you have a global map of all non-store uses, it shouldn't be necessary at all.

zoecarver replied:
I removed this while addressing your below comment.

slavapestov commented on lib/SILOptimizer/IPO/GlobalOpt.cpp:
```diff
@@ -770,6 +786,35 @@ void SILGlobalOpt::optimizeInitializer(SILFunction *AddrF,
+  }
+
+  if (canRemove) {
+    store->eraseFromParent();
```
What if the global is a 'var' and there are multiple stores, but no loads? I'm wondering if it would be better to do this transformation in multiple phases:
- first find all globals that are never loaded
- then remove alloc_global and store instructions
- then remove the globals themselves

zoecarver replied:
I like your idea of multiple phases a lot. I updated this patch, as you suggested, and it is much cleaner/simpler. I have to iron out a few wrinkles; then, I'll push it.

slavapestov reviewed (commented):
This looks great. I have some optional suggestions for further generalizations.

zoecarver committed:
Cleanup

zoecarver commented on lib/SILOptimizer/IPO/GlobalOpt.cpp:
```diff
@@ -955,27 +1026,72 @@ bool SILGlobalOpt::run() {
+  }
+
+  for (auto &global : Module->getSILGlobals()) {
+    // TODO: For some reaons, if we keep iterating through `getSILGlobals`
```
@slavapestov any ideas about the best way to resolve this?

slavapestov replied:
Sounds like an iterator invalidation issue, because you're modifying Module->getSILGlobals() while iterating over it. The idiomatic solution in cases like this is to put the globals that will be removed into a temporary vector, and then iterate over that vector and remove each one.

zoecarver replied:
Yep, that fixed it, thanks!

zoecarver commented on lib/SILOptimizer/IPO/GlobalOpt.cpp:
```diff
@@ -770,6 +798,43 @@ void SILGlobalOpt::optimizeInitializer(SILFunction *AddrF,
+    return false;
+
+  for (auto *addr : GlobalAddrMap[global]) {
+    eraseUsesOfInstruction(addr, [](SILInstruction *) {});
```
Didn't see there's an overload for this. Will remove.

zoecarver committed:
Move collection into collect method

zoecarver commented:
I moved the collection of instructions / global variables into a `collect` method and added a `reset` method. After I erase instructions, I trigger `reset` then `collect` so that the data is up to date, and the subsequent passes can successfully optimize the code. If there is a way to trigger a re-run of the pass, that might be cleaner. Feel free to suggest alternatives. 

I'll finish writing tests tomorrow.

zoecarver committed:
Add tests

zoecarver commented:
Any other review comments (or comments I missed)?

slavapestov commented on lib/SILOptimizer/IPO/GlobalOpt.cpp:
```diff
@@ -59,6 +59,8 @@ class SILGlobalOpt {
 
   typedef SmallVector<ApplyInst *, 4> GlobalInitCalls;
   typedef SmallVector<LoadInst *, 4> GlobalLoads;
+  typedef SmallVector<BeginAccessInst *, 4> GlobalAccess;
```
Maybe "GlobalAccesses" to be consistent with the others which are plural

slavapestov commented on lib/SILOptimizer/IPO/GlobalOpt.cpp:
```diff
@@ -770,6 +804,45 @@ void SILGlobalOpt::optimizeInitializer(SILFunction *AddrF,
+    addr->eraseFromParent();
+  }
+
+  reset();
```
I worry that calling collect() every time something changes is going to turn this pass into an O(n^2) algorithm. Is there a way to build a list of all instructions to be deleted, and then delete them at the end, and finally reset()?

zoecarver replied:
Before deleting globals, I need to recollect everything in order to figure out what's now-dead. This will make it an O(n2) -- is that OK?

slavapestov replied:
That's OK (and it answers one of my questions in the latest review) but my guess would be that it should be possible to determine in one pass both which instructions to delete, and which SIL globals become dead as a result.

zoecarver replied:
Yes, see [my comment](https://github.com/apple/swift/pull/28324#discussion_r356409486).

slavapestov commented on test/SILOptimizer/remove_unused_global_vars_wmo.swift:
```diff
@@ -0,0 +1,27 @@
+// RUN: %target-swift-frontend -O -wmo -emit-sil %s | %FileCheck %s
```
You could probably combine these into one test with two RUN: lines. You can use the --check-prefix flag to FileCheck to change the prefix from CHECK to something else, like CHECK-WMO.

slavapestov reviewed (commented):
Sorry for the delay in responding to this. Thanksgiving week was a holiday for us.

slavapestov commented:
@swift-ci Please smoke test

zoecarver committed:
Merge branch 'master' into optimize/dead-global-vars

zoecarver committed:
Merge branch 'master' into optimize/dead-global-vars

zoecarver commented:
@slavapestov thanks for the review! I'll resolve your comment and look into the build failure.

zoecarver committed:
Create list of instructions to delete

zoecarver committed:
Rename GlobalAccess to GlobalAccesses

zoecarver committed:
Resolve build failure

zoecarver committed:
add helper function: canBeUsedOrChangedExternally

zoecarver commented:
The build failure occurred because a) I didn't exit early in `tryRemoveGlobalAddr` and b) because I didn't check `GlobalVarSkipProcessing`. 

Also, I've updated all uses of `.count(key)` on a map of vectors to now be `[key].size()` because we no longer re-collect between every "optimization," so those elements may exist as empty vectors.

zoecarver committed:
Check size of element, not if the element exists

zoecarver committed:
Remove duplicate test and fix formatting

zoecarver committed:
Merge branch 'master' into optimize/dead-global-vars

zoecarver committed:
stash

zoecarver committed:
Fix invalid global addr being collected

zoecarver committed:
Add extra safety check before removing globals and add accessors even in the global accessor function

zoecarver committed:
Check that the only uses of a global addrs are stores

zoecarver committed:
Remove debug dump call

zoecarver commented:
@slavapestov I think I've resolved all the build issues (builds successfully on my machine). Can you re-trigger the bot?

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please benchmark

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented on lib/SILOptimizer/IPO/GlobalOpt.cpp:
```diff
@@ -808,6 +857,50 @@ static bool canBeChangedExternally(SILGlobalVariable *SILG) {
 }
 
+static bool canBeUsedOrChangedExternally(SILGlobalVariable *global) {
+  if (global->isLet())
```
When does this produce a different result in the isLet() case?

zoecarver replied:
Looking at it, I think they should always be the same but this way it will always match [the check here](https://github.com/apple/swift/blob/eb39989f10c1ca70e0542c65230238c1a7a07834/lib/SILOptimizer/IPO/GlobalOpt.cpp#L932-L937).

slavapestov commented on lib/SILOptimizer/IPO/GlobalOpt.cpp:
```diff
@@ -808,6 +857,50 @@ static bool canBeChangedExternally(SILGlobalVariable *SILG) {
+      AllocGlobalStore.count(global) || GlobalVarStore.count(global))
+    return false;
+
+  Module->eraseGlobalVariable(global);
```
Can this add SILGlobalVariables to delete to a vector as well? This will prevent having to create a vector that stores _all_ global variables...

slavapestov commented on lib/SILOptimizer/IPO/GlobalOpt.cpp:
```diff
@@ -937,13 +1043,31 @@ void SILGlobalOpt::optimizeGlobalAccess(SILGlobalVariable *SILG,
-    // Don't optimize functions that are marked with the opt.never attribute.
-    if (!F.shouldOptimize())
-      continue;
+void SILGlobalOpt::collectUsesOfInstructionForDeletion(SILInstruction *inst) {
```
There might be a utility for doing this already. The safest way is to use a worklist, instead of a recursive function, in case you come across cycles or deep paths in the def-use graph.

zoecarver replied:
Good to know, I'll use a worklist.

zoecarver replied:
I updated the code to use a worklist but I don't think it works well here. We can't iterate over a worklist which is necessary in `tryRemoveGlobalAddr`. I'm going to change it back to a `SmallVector` and use `eraseUsesOfInstruction` when removing instructions.

slavapestov commented on lib/SILOptimizer/IPO/GlobalOpt.cpp:
```diff
@@ -966,27 +1090,104 @@ bool SILGlobalOpt::run() {
+  // Copy the globals so we don't get issues with modifying while iterating.
+  SmallVector<SILGlobalVariable *, 12> globals;
+  for (auto &global : Module->getSILGlobals()) {
+    globals.push_back(&global);
```
This shouldn't be necessary if tryRemoveUnusedGlobal() instead adds the globals to remove onto a new list.

zoecarver replied:
Good catch.

slavapestov commented on lib/SILOptimizer/IPO/GlobalOpt.cpp:
```diff
@@ -966,27 +1090,104 @@ bool SILGlobalOpt::run() {
+
+  // After we erase some instructions, re-collect.
+  reset();
+  collect();
```
If the second call to collect() is just in the service of tryRemoveUnusedGlobal(), it would be wiser to split up the analysis into two parts, with the second part only running after instructions have been erased, as above.

zoecarver replied:
`tryRemoveUnusedGlobal()` checks `GlobalAddrMap`, `GlobalAccessMap`, `GlobalLoadMap`, etc. so these need to be re-collected. Otherwise, it wouldn't be able to optimize away a global variable in one pass.

If SILGlobalVariable had a `getUses` method, we wouldn't need to recollect. I'm not sure if there is a reason it doesn't, but, I'll look into making a patch to add one.

zoecarver replied:
Alternatively, I could check that the global isn't in those maps (`GlobalAddrMap`, `GlobalAccessMap`, `GlobalLoadMap`, etc.) or that all values (corresponding to the global) are an instruction in `InstToRemove`. It would be a messier but probably faster.

slavapestov replied:
The last approach is what I had in mind. You might find `std::all_of()` helpful to collapse the innermost loop.

slavapestov commented on lib/SILOptimizer/IPO/GlobalOpt.cpp:
```diff
@@ -966,27 +1090,104 @@ bool SILGlobalOpt::run() {
+  }
+
+  // Reset incase we re-run this function (when HasChanged is true).
+  reset();
```
Another typo: "incase"

slavapestov commented:
@swift-ci Please benchmark

zoecarver commented:
Arg... more failures. I'll look into those tomorrow.

slavapestov commented:
```
01:10:49 Failing Tests (14):
01:10:49     Swift(linux-x86_64) :: stdlib/String.swift
01:10:49     Swift(linux-x86_64) :: stdlib/CharacterPropertiesLong.swift
01:10:49     Swift(linux-x86_64) :: stdlib/Bool.swift
01:10:49     Swift(linux-x86_64) :: stdlib/VarArgs.swift
01:10:49     Swift(linux-x86_64) :: stdlib/StringAPI.swift
01:10:49     Swift(linux-x86_64) :: stdlib/RawRepresentable-tricky-hashing.swift
01:10:49     Swift(linux-x86_64) :: stdlib/Mirror.swift
01:10:49     Swift(linux-x86_64) :: stdlib/Character.swift
01:10:49     Swift(linux-x86_64) :: stdlib/CharacterProperties.swift
01:10:49     Swift(linux-x86_64) :: SILOptimizer/static_strings.swift
01:10:49     Swift(linux-x86_64) :: SILGen/global_resilience.swift
01:10:49     Swift(linux-x86_64) :: SIL/Serialization/perf_inline_without_inline_all.swift
01:10:49     Swift(linux-x86_64) :: Interpreter/extended_grapheme_cluster_literal.swift
01:10:49     Swift(linux-x86_64) :: DebugInfo/inlinescopes.swift
```

It looks like a few of the failures are actually because a few tests that were matching the generated SIL now have some of their global variables optimized out. Depending on the test it either makes sense to update the expected SIL, or change the code somehow to prevent the global from being optimized out.

On the other hand, a few execution tests are failing with this:

```
00:15:44 Fatal error: file /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/swift/stdlib/public/core/BridgeObjectiveC.swift, line 126
00:15:44 Current stack trace:
00:15:44 0    libswiftCore.dylib                 0x0000000106f982c0 _swift_stdlib_reportFatalErrorInFile + 116
00:15:44 1    libswiftCore.dylib                 0x0000000106c12030 closure #1 in closure #1 in closure #1 in _assertionFailure(_:_:file:line:flags:) + 519
00:15:44 2    libswiftCore.dylib                 0x0000000106c11ca0 closure #1 in closure #1 in _assertionFailure(_:_:file:line:flags:) + 87
00:15:44 3    libswiftCore.dylib                 0x0000000106c11900 closure #1 in _assertionFailure(_:_:file:line:flags:) + 99
00:15:44 4    libswiftCore.dylib                 0x0000000106c11280 _assertionFailure(_:_:file:line:flags:) + 561
00:15:44 5    libswiftCore.dylib                 0x0000000106c117b0 _fatalErrorMessage(_:_:file:line:flags:) + 32
00:15:44 6    libswiftCore.dylib                 0x0000000106ca1b50 String._bridgeToObjectiveCImpl() + 94
00:15:44 7    libswiftFoundation.dylib           0x0000000107883910 StringProtocol.range<A>(of:options:range:locale:) + 360
00:15:44 8    libswiftStdlibUnittest.dylib       0x0000000107c8e0e0 findSubstring(_:_:) + 312
00:15:44 9    libswiftStdlibUnittest.dylib       0x0000000107c8b9f0 processLine #1 (_:isStdout:) in _ParentProcess._runTestInChild(_:_:parameter:) + 281
00:15:44 10   libswiftStdlibUnittest.dylib       0x0000000107c8a280 _ParentProcess._runTestInChild(_:_:parameter:) + 4275
00:15:44 11   libswiftStdlibUnittest.dylib       0x0000000107c8e320 _ParentProcess.runOneTest(fullTestName:testSuite:test:testParameter:) + 2455
00:15:44 12   libswiftStdlibUnittest.dylib       0x0000000107c90720 _ParentProcess.run() + 4276
00:15:44 13   libswiftStdlibUnittest.dylib       0x0000000107c93640 runAllTests() + 1965
00:15:44 14   TestTimeZone                       0x0000000106baca80 main + 1455
00:15:44 15   libdyld.dylib                      0x00007fff7a9053d4 start + 1
00:15:44 /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/buildbot_incremental/swift-macosx-x86_64/test-macosx-x86_64/stdlib/Output/TestTimeZone.swift.script: line 5: 95420 Illegal instruction: 4  /usr/bin/env DYLD_LIBRARY_PATH='/Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/buildbot_incremental/swift-macosx-x86_64/lib/swift/macosx' LD_LIBRARY_PATH='/Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/buildbot_incremental/swift-macosx-x86_64/lib/swift/macosx:' SIMCTL_CHILD_DYLD_LIBRARY_PATH='/Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/buildbot_incremental/swift-macosx-x86_64/lib/swift/macosx' /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/buildbot_incremental/swift-macosx-x86_64/test-macosx-x86_64/stdlib/Output/TestTimeZone.swift.tmp/TestTimeZone > /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/buildbot_incremental/swift-macosx-x86_64/test-macosx-x86_64/stdlib/Output/TestTimeZone.swift.tmp.txt
```

It looks like a miscompile. I would check that you're not accidentally deleting live instructions. I would probably first look at the recursive walk of the def-use graph in `collectUsesOfInstructionForDeletion()` and add some assertions there to see that we're not deleting anything we shouldn't be.

zoecarver commented:
@slavapestov thanks for the analysis! The good news is there are only 14 tests failing on Linux, the bad news is there are 645 failing on OS X. 

I'm trying to reproduce one of the miscompiles but I can't seem to be able to. Even running with lit, I can't get `sdk-link.swift`, `TestTimeZone.swift`, or `enum-error-execute.swift` to fail. Any ideas why? Maybe some config or my OS version (10.14.5)?

zoecarver committed:
Fix inlinescopes test

zoecarver committed:
Check global can be used before removing alloc

zoecarver committed:
Add assertion in collectUsesOfInstructionForDeletion

zoecarver committed:
* use worklist instead of vector
* replace dead global collection with global vector
* fix test

zoecarver committed:
Fix use of worklist

zoecarver committed:
Remove double-collect

zoecarver committed:
Fix issues with worklist

zoecarver committed:
Fix typo

zoecarver commented:
@slavapestov I've addressed your comments and fixed the tests some of the tests. While debugging some of the tests, I discovered that `tryRemoveGlobalAlloc` didn't check `isSafeToRemove` which meant that sometimes globals would be used but not allocated. I suspect that is the cause of most of the miscompiles (which would explain why the swift program crashed and not the compiler). 

I'll try to run all the tests tonight and fix the remaining failing tests.

zoecarver committed:
Fix logic around checking dead instructions

zoecarver committed:
Add comments and fix checking dead instructions

zoecarver commented on lib/SILOptimizer/IPO/GlobalOpt.cpp:
```diff
@@ -20,6 +20,7 @@
 #include "swift/SIL/SILCloner.h"
 #include "swift/SIL/SILGlobalVariable.h"
 #include "swift/SIL/SILInstruction.h"
+#include "swift/SIL/SILInstructionWorklist.h"
```
This is no longer needed.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please benchmark

zoecarver committed:
Merge branch 'master' into optimize/dead-global-vars

zoecarver commented:
It looks like this patch caused a regression in `static_strings.swift` does that make sense or am I misdiagnosing the issue?

theblixguy commented:
Maybe... it could be just that the test needs to be updated because the global is optimised out? The test is not supported on macOS so it was skipped.

(Also, do you mind squashing your commits? Patch has grown to nearly 30 commits).

eeckstein commented:
@zoecarver it looks like a real regression. Can you please investigate this before merging?
@theblixguy The test passes on macOS, it's not disabled.

theblixguy commented:
Oh, CI logs said “UNSUPPORTED” so I assumed it was skipped.

eeckstein commented:
Wait, you are right. This test requires the stdlib to be built in no-assert mode. I think we are doing this only on linux. So it's in fact not running on our macOS CI bot. But it's not disabled for macOS in general.

zoecarver commented:
Alrighty, this will be fun to debug :) I suspect it has to do with the modified collection of globals. 

@theblixguy I'll rebase after I fix the regression.

zoecarver committed:
Fix should optimize logic

zoecarver committed:
Update requirements for test

zoecarver commented on test/SILOptimizer/static_strings.swift:
```diff
@@ -5,7 +5,7 @@
 // RUN: %target-run %t/a.out | %FileCheck %s -check-prefix=CHECK-OUTPUT
 
-// REQUIRES: executable_test,swift_stdlib_no_asserts,optimized_stdlib
+// REQUIRES: executable_test,swift_stdlib_no_asserts
```
@eeckstein git blame showed that you were the one who added this requirement. The test passes without `optimized_stdlib` do you think this is an OK change?

zoecarver commented:
I finally was able to reproduce/fix the issue. Thanks for the patience. Could someone re-trigger the bots?

slavapestov commented:
Sorry, we were all out on vacation, again.

@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

zoecarver commented:
Friendly ping. Does this need any more reviewing or can it be merged?
